src/Runner/TestSuiteSorter.php

```diff
@@ -31,27 +32,74 @@
      */
     public const ORDER_REVERSED = 2;
 
+    /**
+     * @var int
+     */
+    public const ORDER_DEFECTS_FIRST = 3;
+
+    /**
+     * @var array
+     */
+    private $defectSortWeight = [
+        ResultCacheListener::TEST_SKIPPED    => 7,
+        ResultCacheListener::TEST_ERROR      => 6,
+        ResultCacheListener::TEST_FAILURE    => 5,
+        ResultCacheListener::TEST_WARNING    => 4,
+        ResultCacheListener::TEST_INCOMPLETE => 3,
+        ResultCacheListener::TEST_RISKY      => 2,
+        ResultCacheListener::TEST_UNKNOWN    => 0
+    ];
+
+    /**
+     * @var array
+     */
+    private $defectSortOrder;
+
+    /**
+     * @var TestResultCache
+     */
+    private $cache;
+
     /**
      * @throws Exception
      */
-    public function reorderTestsInSuite(Test $suite, int $order, bool $resolveDependencies): void
+    public function reorderTestsInSuite(Test $suite, int $order, bool $resolveDependencies, int $orderDefects): void
     {
         if ($order !== self::ORDER_DEFAULT && $order !== self::ORDER_REVERSED && $order !== self::ORDER_RANDOMIZED) {
             throw new Exception(
                 '$order must be one of TestSuiteSorter::ORDER_DEFAULT, TestSuiteSorter::ORDER_REVERSED, or TestSuiteSorter::ORDER_RANDOMIZED'
             );
         }
 
-        if ($suite instanceof TestSuite && !empty($suite->tests())) {
+        if ($orderDefects !== self::ORDER_DEFAULT && $orderDefects !== self::ORDER_DEFECTS_FIRST) {
+            throw new Exception(
+                '$orderDefects must be one of TestSuiteSorter::ORDER_DEFAULT, TestSuiteSorter::ORDER_DEFECTS_FIRST'
+            );
+        }
+
+        if ($orderDefects === self::ORDER_DEFECTS_FIRST && $this->cache === null) {
+            $this->cache = new TestResultCache();
+            $this->cache->loadFromFile();
+        }
+
+        if ($suite instanceof TestSuite) {
             foreach ($suite as $_suite) {
-                $this->reorderTestsInSuite($_suite, $order, $resolveDependencies);
+                $this->reorderTestsInSuite($_suite, $order, $resolveDependencies, $orderDefects);
             }
 
-            $this->sort($suite, $order, $resolveDependencies);
+            if ($orderDefects === self::ORDER_DEFECTS_FIRST) {
+                $this->addSuiteToDefectSortOrder($suite);
+            }
+            $this->sort($suite, $order, $resolveDependencies, $orderDefects);
         }
     }
 
-    private function sort(TestSuite $suite, int $order, bool $resolveDependencies): void
+    public function setCache(TestResultCache $cache): void
```

Comment:
I believe this property shall not be exposed

Reply:
You're right, the method is only used in the test. Will clean it up :)